website/docs/optimizations/java/optimized-jvm.md

```diff
@@ -0,0 +1,154 @@
+:::
+
+Start the build for the container image. While it is building, you can move to the next step and inspect the Dockerfile.
+
+```bash showLineNumbers
+cd ~/environment/unicorn-store-spring
+docker buildx build --load -t unicorn-store-spring:latest .
+IMAGE_TAG=i$(date +%Y%m%d%H%M%S)
+docker tag unicorn-store-spring:latest $ECR_URI:$IMAGE_TAG
+docker tag unicorn-store-spring:latest $ECR_URI:latest
+docker push $ECR_URI:$IMAGE_TAG
+docker push $ECR_URI:latest
+```
+
+Inspect the modified `Dockerfile`.
+
+```dockerfile showLineNumbers {11-14,20-23,28-29}
+FROM public.ecr.aws/docker/library/maven:3.9-amazoncorretto-17-al2023 as builder
+
+RUN yum install -y tar gzip unzip
+
+COPY ./pom.xml ./pom.xml
+RUN mvn dependency:go-offline -f ./pom.xml
+
+COPY src ./src/
+RUN mvn clean package && mv target/store-spring-1.0.0-exec.jar target/store-spring.jar && cd target && unzip store-spring.jar
+
+RUN jdeps --ignore-missing-deps \
+    --multi-release 17 --print-module-deps \
+    --class-path="target/BOOT-INF/lib/*" \
+    target/store-spring.jar > jre-deps.info
+
+# Adding jdk.crypto.ec for TLS 1.3 support
+RUN truncate --size -1 jre-deps.info
+RUN echo ",jdk.crypto.ec" >> jre-deps.info && cat jre-deps.info
+
+RUN export JAVA_TOOL_OPTIONS=\"-Djdk.lang.Process.launchMechanism=vfork\" && \
+    jlink --verbose --compress 2 --strip-java-debug-attributes \
+    --no-header-files --no-man-pages --output custom-jre \
+    --add-modules $(cat jre-deps.info)
+
+FROM public.ecr.aws/amazonlinux/amazonlinux:2023.2.20231026.0
+RUN yum install -y shadow-utils
+
+COPY --from=builder target/store-spring.jar store-spring.jar
+COPY --from=builder custom-jre custom-jre
+
+RUN groupadd --system spring -g 1000
+RUN adduser spring -u 1000 -g 1000
+
+USER 1000:1000
+
+# OpenTelemetry agent configuration
+ENV OTEL_TRACES_SAMPLER "always_on"
+ENV OTEL_PROPAGATORS "tracecontext,baggage,xray"
+ENV OTEL_RESOURCE_ATTRIBUTES "service.name=unicorn-store-spring"
+ENV OTEL_IMR_EXPORT_INTERVAL "10000"
+ENV OTEL_EXPORTER_OTLP_ENDPOINT "http://localhost:4317"
+
+EXPOSE 8080
+ENTRYPOINT ["./custom-jre/bin/java","-jar","-Dserver.port=8080","/store-spring.jar"]
+```
+
+As discussed in the first chapter the Dockerfile leverages a multi-stage build. In the first stage (starting line 1), the application and a custom runtime is built. The custom runtime is based on the list of dependencies that have been collected with `jdeps` (see line 11 - 14).
+
+In the first step, the complete classpath of the application is analyzed and all module dependencies are written in a file called `jre-deps.info`. You may have noticed the explicit addition of `jdk.crypto.ec`. This module provides the implementation of the SunEC security provider which is necessary for TLS support. Unfortunately, it is not obvious from the analysis of the classpath that this additional module is necessary. For this reason, it is added at this point.
```

Comment:
In the first step, we analyze the entire classpath of the application and write down all module dependencies in a file called `jre-deps.info`. You might have noticed that we explicitly added `jdk.crypto.ec`. This module contains the implementation of the SunEC security provider, which is essential for TLS support. However, it is not easy to determine from the classpath analysis that this module is needed, which is why we include it at this stage.